service/src/server/routes/subgraphs.rs

```diff
@@ -124,7 +127,13 @@ pub async fn subgraph_queries(
                 metrics::FAILED_QUERIES
                     .with_label_values(&[&deployment_label])
                     .inc();
-                bad_request_response("Bad response from Graph node")
+                IndexerError::new(
+                    IndexerErrorCode::IE032,
+                    Some(IndexerErrorCause::new(
+                        "Failed to execute a paid subgraph query to graph node",
+                    )),
```

Comment:
Looking at this, where we essentially drop the error returned by Graph Node and just log a static string makes me wonder if we should have an `IndexerErrorCause` that supports taking other errors (e.g. the `std::error::Error` trait) that are then included in the error message.

The TypeScript indexer-service/-common does this by having `IndexerErrorCause` be an arbitraty type. Easy in TypeScript of course. Here, what we could do is something like
```rust
use std::fmt::{self};
use anyhow::anyhow; // 1.0.75

#[derive(Debug)]
enum IndexerErrorCode {
    IE001,
}

struct IndexerErrorCause<T: ToString>(T);

impl<T: ToString> From<T> for IndexerErrorCause<T> {
    fn from(t: T) -> Self {
        Self(t)
    }
}

impl<T: ToString> fmt::Display for IndexerErrorCause<T> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.0.to_string())
    }
}

struct IndexerError<T: ToString> {
    code: IndexerErrorCode,
    cause: IndexerErrorCause<T>
}

impl<T> IndexerError<T>
where T: ToString + Into<IndexerErrorCause<T>>
{
    pub fn new(code: IndexerErrorCode, cause: T) -> Self {
        Self { code, cause: cause.into() }
    }
}

impl<T: ToString> fmt::Display for IndexerError<T> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}: {}", self.code, self.cause.to_string())
    }
}

fn main() {
    println!("{}", IndexerError::new(IndexerErrorCode::IE001, "str cause"));
    println!("{}", IndexerError::new(IndexerErrorCode::IE001, anyhow!("anyhow cause")));
}
```
You can see this in the playground here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5ff50fb1ffb9afe9574f1087ebb581fa

Reply:
Using `Display` instead of `ToString` might be better, I'm not sure which one is more common in situations like this. Perhaps @Theodus knows?

Reply:
`ToString` is generally preferred. When you implement `ToString`, you get the impl of `Display` automatically ([docs](https://doc.rust-lang.org/std/string/trait.ToString.html))

Reply:
But it seems like an alternative would be `cause: anyhow::Error`. Since it looks like you're recreating a subset of what `anyhow::Error` does, but with the additional overhead of the error type not getting erased. Unless you want the type for some reason.

Reply:
Thanks for the detailed review here 🧠 !!

Looking at the TS implementation of subgraph query [responses](https://github.com/graphprotocol/indexer/blob/a73b7dd41791e2ea44a45f79641939e687434abf/packages/indexer-service/src/server/index.ts#L311), I realized that we have been incorrectly only counting `200` response from graph node as Successful queries, whereas TS counts all responses when the request doesn't error out (It seems like to me). 

When we get error for executing the query, we get `QueryError` which can be passed into `IndexerErrorCause` directly as `IndexerErrorCause` is currently defined as `pub struct IndexerErrorCause(Box<dyn Error + Send + Sync>);`. 
Updating it to `anyhow::Error` could be nice, but I think the current struct accomplishes what we want to achieve here.

For now, I updated in [e56b400](https://github.com/graphprotocol/indexer-rs/pull/79/commits/e56b400f09ea83432f5a79e91df1f35e6629e4d9) so the match statement will count `FAILED_QUERIES` only if `execute_paid_query` fails and it would log out and respond with "Failed to execute a paid subgraph query to graph node: " + `QueryError` back to the user

lmk what you think:)